pkg/network/dns_cache.go

```diff
@@ -71,17 +68,16 @@ func (c *reverseDNSCache) Add(translation *translation, now time.Time) bool {
 		return false
 	}
 
-	exp := now.Add(c.ttl).UnixNano()
-	for _, addr := range translation.ips {
+	for addr, deadline := range translation.ips {
 		val, ok := c.data[addr]
 		if ok {
-			val.expiration = exp
-			if rejected := val.merge(translation.dns, c.maxDomainsPerIP); rejected && c.oversizedLogLimit.ShouldLog() {
+			if rejected := val.merge(translation.dns, deadline, c.maxDomainsPerIP); rejected && c.oversizedLogLimit.ShouldLog() {
 				log.Warnf("%s mapped to too many domains, DNS information will be dropped (this will be logged the first 10 times, and then at most every 10 minutes)", addr)
 			}
 		} else {
 			atomic.AddInt64(&c.added, 1)
-			c.data[addr] = &dnsCacheVal{names: []string{translation.dns}, expiration: exp}
+			// flag as in use, so mapping survives until next time connections are queried, in case TTL is shorter
+			c.data[addr] = &dnsCacheVal{names: map[string]time.Time{translation.dns: deadline}, inUse: true}
```

Comment:
Should this be protected by a call to `c.mux.Lock()` ?

Reply:
The mux is already locked at this point, see line 65.

Reply:
My bad, I missed that!